### ISSUE TITLE
[Feature Request] Embed dashboards as Read-only dashboard

### ISSUE THREAD

**Issue by nook24** [grafana/grafana#13706]
Read before posting: 

- Questions should be posted to https://community.grafana.com. Please search there and here on GitHub for similar issues before creating a new issue. 
- Checkout FAQ: https://community.grafana.com/c/howto/faq
- Checkout How to troubleshoot metric query issues: https://community.grafana.com/t/how-to-troubleshoot-metric-query-issues/50

Please include this information:

**Comment by nook24** [comment]
@torkelo Security by obscurity is not a good good idea at all. However, I'm not looking for an bulletproof method, otherwise I would split up dashboards into different grafana instances or so.

I have created an basic editor, where Users can create own Grafana dashboards. In this editor, the user can only add its own objects to the Dashboard. (Linux admins see Linux-Servers, Windows admins only see Windows-Servers, etc.). So they don't need to mess around with the used data source, writing queries and so on.
![bildschirmfoto 2018-10-17 um 14 09 33](https://user-images.githubusercontent.com/9019992/47085671-9921a580-d217-11e8-84e8-9ab288e39cea.png)

The system will automatically generate a Grafana dashboard through the API, which is embedded using an `<iframe src="...&kiosk">`
![bildschirmfoto 2018-10-17 um 14 05 53](https://user-images.githubusercontent.com/9019992/47085872-41d00500-d218-11e8-931c-e4cc5e71301b.png)

If I press "esc" or "F" inside the iframe, I get dropped out of the kiosk mode and get the list of all dashboards:
![bildschirmfoto 2018-10-17 um 14 06 25](https://user-images.githubusercontent.com/9019992/47086083-dd617580-d218-11e8-9909-85c8628814f5.png)
This is exactly what i try to avoid for non technical users. It should not be possible for them to "break out" of the current dashboard, by pressing a few random keys.

It is ok for me, if a user copy&past the url out of the iframe and can access Grafana itself. Otherwise it would not run with anonymous mode enabled. 

--- 
I guess, if I try only embed the panels of the dashboard, this will terminate responsive support, because every iframe needs a fixed height and width (as far as i know?)

**Comment by egalecki** [comment]
Hi everyone,
I encounter the same case. Is there any news about a feature like this ?

**Comment by nehaJayadeep** [comment]
Hi any update on this. I am also looking for this feature where the user can access only specific graphs through link. A  Way to disable short cut keys should solve this problem i suppose. Please help.

**Comment by dannielwallace** [comment]
Hi any update on this. I am also looking for this feature where the user can access only specific graphs through link. A Way to disable short cut keys should solve this problem i suppose. Please help.

**Comment by pemartini** [comment]
Any updates?

**Comment by baur** [comment]
also looking for this feature

**Comment by ayamshanov** [comment]
...and me too.

**Comment by egalecki** [comment]
Hi everyone,

Due to non response from grafana, I solved my problem by using the ELK stack which allows you to share dashboards in iframes whithout this problem.

You can use logstash to get data from Mysql for example by using a jdbc plugin (everything is explained on logstash documentation) and store them in an elasticsearch node to make dashboards in kibana.

I hope this message could help.

**Comment by ctaity** [comment]
No news on this? I need too.

**Comment by MikPak** [comment]
This feature would be great.

**Comment by Ermag** [comment]
If you embed Grafana from the same host (reverse proxy) you can access the iframe with JS and run this snippet inside it.
`IFRAME.contentWindow.Mousetrap.unbindGlobal('esc');
IFRAME.contentWindow.Mousetrap.unbind('esc')`

**Comment by laundmo** [comment]
i wanted to explain how to set up the solution by @Ermag with permissions that block access to otther dashboards

I managed to embed grafana (from another host) after following https://grafana.com/docs/auth/auth-proxy/ 
step by step:
1. edit grafana config accordingly to docs page
2. add apache server ip to whitelist parameter (all traffic from this ip has access to all accoutns without password)
3. edit apache config according to docs page
4. `ServerAdmin`, `ServerName`, `ErrorLog`, `CustomLog` can be omitted if not needed
5. edit `ProxyPass` and `ProxyPassReverse` to grafana url/ip and port
6. add a user with the same name as the grafana user to htpasswd
7. edit permissions of that user to Viewer for only the appropriate dashboard in grafana (to make sure the user can only view the correct dashboard)
8. embed `<iframe src="http(s)://username:password@your-grafana-url.com?kiosk"` (make sure your page has other authentiation before this embed to prevent unauthorized access to clear text username:password)
9. optional: Add JS that stops pressing ESC according to Ermag

Now you can create as many users as you want and add them to both grafana and htpasswd (IMPORTANT same username) and manage their permissions in grafana while embeding their dashboards however you want.
i STRONGLY suggest using https since the nthe clear text password is encrypted while going through network, insted of http where everyone can listen in and get the password.

**Comment by MasiGit** [comment]
Any updates on this feature?

**Comment by zixiliuyue** [comment]
This file [keybindingSrv.ts]( https://github.com/grafana/grafana/blob/master/public/app/core/services/keybindingSrv.ts), I deleted these 
![image](https://user-images.githubusercontent.com/12575317/76045607-994f4c00-5f98-11ea-83c2-7378879026d9.png) . Then recompile the front end [build](https://github.com/grafana/grafana/blob/master/contribute/developer-guide.md). upload /usr/share/grafana/public or homepath.

**Comment by ojathelonius** [comment]
For anyone willing to prevent such behavior without having to rebuild the front-end, here's a hacky solution :

```
$('#your-grafana-iframe')[0].contentDocument.addEventListener('keydown', function (event) {
    this.getElementsByTagName('body')[0].classList.add('view-mode--kiosk');
});
```

Basically you add the class that is being removed when pressing the Esc key. Works for me !

**Comment by rubn-g** [comment]
I would also love to have this feature, it's very helpful to show a dashboard integrated in an iframe, and preventing the user from changing the dashboard ui just by pressing a key by mistake. Maybe we could have an extra parameter after &kiosk (like &static, or whatever), that disables the key events?

**Comment by EdisonSu768** [comment]
> If you embed Grafana from the same host (reverse proxy for example) you can access the iframe with JS and run this snippet to unbind the keypress events:
> `IFRAME.contentWindow.Mousetrap.unbindGlobal('esc'); IFRAME.contentWindow.Mousetrap.unbind('esc');`
> 
> I've tested it on version 6.0.1 but i suspect it will work on lower as well.

what about other shortcuts like 'F' for searching, also unbind?

**Comment by yogendra-prasad** [comment]
Thanks, zixiliuyue for the solution, This code changes worked for me. 

Along with this to make the Iframe accessible only from my specific site, We added a response header in middleware as:

	**w.Header().Add("Content-Security-Policy", "frame-ancestors  siteUrl")**. This restricted any other site to embed Grafana url as  iframe.

**Comment by asnowfox** [comment]
<template>
    <div class="main-views">
        <iframe class="frame-class" scrolling="no" ref="grafana" marginwidth="0" marginheight="0" frameborder="0"
            :src="iframeUrl">
        </iframe>
    </div>
</template>

///////js code///////
mounted () {
            this.$refs.grafana.contentWindow.addEventListener('keydown', function (event) {
                if (event.keyCode === 27) {
                    event.preventDefault()
                    event.stopPropagation()
                    return false
                }
            }, true)
        },

in VUE, It works well for my project. hope helpful for you.

**Comment by p4block** [comment]
@asnowfox Thanks a million

I hosted grafana on the same domain as our website (on a subdir) and set this up:

```

**Comment by p4block** [comment]
@asnowfox Thanks a million

I hosted grafana on the same domain as our website (on a subdir) and set this up:

```
<iframe
  id="grafanaframe"
  src="url"
  width="100%"
  height="100%"
  frameborder="0"
></iframe>

<script>
  $("#grafanaframe").on("load", function () {
    this.contentWindow.addEventListener(
      "keydown",
      function (event) {
        if (event.keyCode === 27) {
          event.preventDefault();
          event.stopPropagation();
          return false;
        }
      },
      true
    );
  });
</script>
```

**Comment by nehaJayadeep** [comment]
Yes, that's one way. Even of in different domains this can be achieved by
configuring nginx.

Thanks,
Neha Jayadeep

**Comment by samuelting** [comment]
> @asnowfox Thanks a million
> 
> I hosted grafana on the same domain as our website (on a subdir) and set this up:
> 
> ```
> <iframe
>   id="grafanaframe"
>   src="url"
>   width="100%"
>   height="100%"
>   frameborder="0"
> ></iframe>
> 
> <script>
>   $("#grafanaframe").on("load", function () {
>     this.contentWindow.addEventListener(
>       "keydown",
>       function (event) {
>         // Block user from exiting grafana ui easily (esc and F keys)
>         if (event.keyCode === 27 || event.keyCode === 70 ) {
>           event.preventDefault();
>           event.stopPropagation();
>           return false;
>         }
>       },
>       true
>     );
>   });
> </script>
> ```
> 
> Still would love this feature to exist without js hacking. Grafana is an untapped golden mine for data analysis.

Hello @p4block , 

I think I can come up with a better solution.
Note that Grafana dashboards responds to ESC key strokes only after the containing iframe element receives its focus by the browser. In order to prevent this thing from happening, we can enforce the browser to focus the window / document.body even after it loses focus....

```
<iframe
  id="grafanaframe"
  src="REPLACE_WITH_YOUR_URL"
  width="100%"
  height="100%"
  frameborder="0"
></iframe>

<script>
 window.addEventListener("blur", function (e) {
        setTimeout(function () {
          window.focus();
        }, 0);
      });
</script>
```

This solution do not have interaction with the embedded page. Cross-domain issues can be safely ignored.

**Comment by g-phani-reddy** [comment]
Any updates on this feature?

**Comment by nook24** [comment]
6 years old but still relevant to this day

**Comment by aeppling** [comment]
looking forward for this feature

**Comment by Copilotes** [comment]
+1, very important feather for me.

**Comment by jwavoet** [comment]
We are integrating Grafana into our application to give the end-users more insights. Unfortunatly the end-user can still escape to Grafana itself. We do want to use policies to avoid the end-user making unauthorized changes, however being able to disable 'esc' or any other button combination would be really helpful.

**Comment by Thiago-Dantas** [comment]
what i settled on was embedding public dashboards behind a reverse proxy that only forwards public stuff, so grafana itself is somewhat hidden
not the best, not the worst

**Comment by curtis1757** [comment]
Really need this and to hide panel titles and panel 3 dot menu.....

How hard can this be to implement?????

**Comment by Thiago-Dantas** [comment]
while i would also love this, i kinda solved by going with public dashboards behind nginx
on nginx you only forward the public prefixes and that disallows navigation

**Comment by JagadishaIncture** [comment]
+1

**Comment by hugotms** [comment]
@Thiago-Dantas thanks for the idea ! Unfortunately, this won't work for me as I need my users to be able to use template variables and cannot manage one dashboard per device monitored.

**Comment by kale1d0code** [comment]
You could make an intermediary iFrame (iFrame to page that contains iFrame),
which could have some JavaScript on it which listens for key presses (or add the JS directly to a page if you have the option).
you could make it so nothing happens when esc is pressed.

bit of a faff but works

**Comment by gchiarapa** [comment]
Any updates?

**Comment by m4v3rickmts** [comment]
It’s a widely reported issue, but the developers don’t seem to care enough to fix it or even comment on it.

**Comment by kale1d0code** [comment]
I recommend you don't use grafana in an iFrame

**Comment by curtis1757** [comment]
How to then display grafana dashboard in my webpage?

**Comment by laundmo** [comment]
Grafana has [Externally Shared Dashboards](https://grafana.com/docs/grafana/latest/dashboards/share-dashboards-panels/shared-dashboards/) now which could be a valid option for many usecases. The biggest reason to still iframe the full grafana is: 
- #67346

i recommend upvoting that issue or adding relevant usecases if you have a distinct one from the ones already shared.

**Comment by nook24** [comment]
@laundmo Thanks for mentioning Externally Shared Dashboards, I haven't looked into this for a long time and wasn't aware of this feature. I gave it a quick shot and it's looking promising. 
All my Dashboards are using the `-- Mixed --` datasource. The documentation [explicitly listed Graphite](https://grafana.com/docs/grafana/latest/dashboards/share-dashboards-panels/shared-dashboards/#supported-data-sources) as unsupported data source. For some reason Graphite is working flawlessly in my case, but all panels that use Prometheus are having issues to load the data. I have to do some more digging.

**Comment by silverwind** [comment]
Here's what worked for me:

```ts
iframeDocument.addEventListener("keydown", (e: KeyboardEvent) => {
  if (e.key === "Escape") {
    e.preventDefault();
    e.stopPropagation();
  }
}, {capture: true});

**Comment by ChrisVagg** [comment]
> Here's what worked for me:
> 
> iframeDocument.addEventListener("keydown", (e: KeyboardEvent) => {
>   if (e.key === "Escape") {
>     e.preventDefault();
>     e.stopPropagation();
>   }
> }, {capture: true});

This code can be intercept and change. It is a secure mechanism in the frontend only.

**Comment by silverwind** [comment]
Indeed, a proper solution would need to be in grafana's backend.

**Comment by ChrisVagg** [comment]
i think the best approach is to implement a proxy using [auth.proxy]  or [auth.jwt] with roles. And also implement a reverse proxy and allow only necessary endpoints.

**Comment by hugotms** [comment]
> i think the best approach is to implement a proxy using [auth.proxy] or [auth.jwt] with roles. And also implement a reverse proxy and allow only necessary endpoints.

yes this is what I ended up doing. user can still escape and navigate but the interface just gives them error because the endpoints are blocked.

**Comment by elijahsanders152-stack** [comment]
Yes it is a good option
